bpo-30181: parse docstring using pydoc

brianmay opened the pull request:
Rebased and copied Ben Finney's changes from http://bugs.python.org/issue30181

I hope master is the correct branch to merge this into.

<!-- issue-number: [bpo-30181](https://bugs.python.org/issue30181) -->
https://bugs.python.org/issue30181
<!-- /issue-number -->

benf-wspdigital committed:
Add test cases for stripping surrounding space for shortDescription.

benf-wspdigital committed:
Add test cases for test case docstring with surrounding newlines.

PEP 257 specifies that tools which parse docstrings should strip
surrounding blank lines *before* determining what is the first line
<URL:https://www.python.org/dev/peps/pep-0257/#handling-docstring-indentation>.

benf-wspdigital committed:
Add test cases for single-line test case docstring.

benf-wspdigital committed:
Parse the test case docstring using `pydoc.splitdoc`.

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA. This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://cpython-devguide.readthedocs.io/pullrequest.html#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

brianmay commented:
Apparently trivial changes don't require a CLA. Does this count as a trivial change? In any case, the changes are from Ben Finney, not myself, so I suspect it isn't going to help if I sign the CLA.

brianmay commented:
The two tests that fail don't appear to be valid PEP 257 docstring to me. So I think maybe the tests need to be modified.

serhiy-storchaka commented on Lib/unittest/case.py:
```diff
@@ -338,6 +339,25 @@ def __exit__(self, exc_type, exc_value, tb):
+    """
+    Return the test case short description from a docstring.
+
+    Ths docstring is parsed by the standard `pydoc.splitdoc` function
```
This is an implementation detail. Both `short_description_from_docstring` and `pydoc.splitdoc` are not public API, therefore the docstring is interested only for readers of these sources. It can be just a comment. But there is no much sense to just repeat the code by words in a docstring or comment.

brianmay replied:
There appears to be different ideas at play here surrounding the use docstrings. With one view being docstrings should only document the public API, so users of the API know how to use the package. The other view being docstrings are equally applicable for internal functions, which should have a well defined API for internal use, so internal programmers can tell what the function does without peering into the source. Any external references to help justify one or the other side of this argument appreciated (I have no opinions on this yet myself).

brianmay replied:
I have also had the out of band comment from Ben saying just because it isn't a public API doesn't make the docstring any less useful. Since a docstring will be available to anyone introspecting the object whether private or public

serhiy-storchaka replied:
Docstrings of internal functions are not forbidden, they are just not necessary. Comments would be enough (see PEP 8).

In any case neither comment, nor docstring shouldn't just repeat the code. They should describe *what* function does, but not *how* it does this. Otherwise any change of the code will make a docstring or a comment incorrect.

Note also that the function name itself serves the documenting function.

brianmay replied:
How about:

```python
def short_description_from_docstring(text):
    # Return the summary line from a docstring.
    # If there is no summary line, return None.
    ...
```

The use of the word "summary line" comes straight from PEP-257.

serhiy-storchaka commented on Lib/unittest/test/test_case.py:
```diff
@@ -596,6 +596,48 @@ def testShortDescriptionWithMultiLineDocstring(self):
+    @unittest.skipIf(
+        sys.flags.optimize >= 2,
+        "Docstrings are omitted with -O2 and above")
+    def testShortDescriptionWithSurroundingSpaceOneLineDocstring(self):
```
I think it is enough to just add spaces or newlines in the two preceding tests.

brianmay replied:
Sorry, not sure I understand. You appear to be saying that we should use the same docstring as for the `testShortDescriptionWithOneLineDocstring` test but with extra white space. However it is a different test - doesn't that mean it needs a different doc string to describe how it is different from the previous docstring?

serhiy-storchaka replied:
I mean that adding spaces at the start and end of the docstring of testShortDescriptionWithOneLineDocstring and adding a newline and indentation at the start of the docstring of testShortDescriptionWithMultiLineDocstring will cover cases tested by new testes. No new tests will be needed.

brianmay replied:
Oh, OK, I think I understand now. So to clarify, you feel there is no need to have a separate test for the no-leading/trailing whitespace case?

brettcannon commented:
FYI Ben Finney has signed the CLA, although this is not a trivial change if you make any changes to the PR yourself, @brianmay then we will need you to sign the CLA, so please do so.

serhiy-storchaka commented on Lib/unittest/case.py:
```diff
@@ -463,15 +483,8 @@ def defaultTestResult(self):
-        doc = self._testMethodDoc
-        return doc and doc.split("\n")[0].strip() or None
-
+        """ Return a one-line description of the test, if any; otherwise None. """
```
`shortDescription()` is a public API. Your change removes useful information from its docstring.

brianmay replied:
Not sure that the previous description is entirely relevant. Also does saying "The default implementation" add any useful information? How about:

```
def shortDescription(self):
    """Returns a one-line description of the test, or None if no
    description has been provided.

    This method returns the summary line of
    the specified test method's docstring, as per PEP-257.
    """
    ...
```

brianmay commented:
I have signed the CLA - is it possible to remove "CLA not signed" tag?

brettcannon commented:
Ah, I see the issue. Your PR has code from https://github.com/benf-wspdigital who has not signed the CLA, so the bot is blocking it as the provenance of the code isn't CLA-clean. Any way of getting that git user to sign the CLA?

brianmay commented:
I am confused; that looks like Ben Finney's account, and you previously said he had signed the CLA.

encukou commented:
The bot doesn't recognize him because he has not added his Github username to his [profile at bugs.python.org](https://bugs.python.org/user26026).

brettcannon commented:
@encukou is right. Ben Finney has signed based on his bugs.python.org account, but the bot nor I can verify that benf-swpdigital on GitHub is Ben Finney.

brianmay commented:
https://bugs.python.org/msg293522

encukou commented:
> Ben Finney has signed based on his bugs.python.org account, but the bot nor I can verify that benf-swpdigital on GitHub is Ben Finney.

GitHub determines attribution by the e-mail in the commit. This is just text, and easily spoofable – notice how the commits here ended up attributed to Ben without any action on his part, and without GitHub's knowledge of what's on b.p.o.
So, you can tell `benf-swpdigital` on GitHub registered with the e-mail address in these commits.

Anyway, instead you probably want to verify the patches here are the ones from Ben, i.e. compare them to the ones on b.p.o.

brettcannon commented:
Just an FYI, this whole CLA issue of people being nice and opening PRs on other people's behalf is being discussed at python/devguide#195 .

terryjreedy commented:
There would have been less confusion about who has not signed the CLA if the Knight's explanatory message began with "Hello <git-name>:".

brettcannon commented:
@terryjreedy python/the-knights-who-say-ni#42

cjerdonek commented on Lib/unittest/case.py:
```diff
@@ -338,6 +339,25 @@ def __exit__(self, exc_type, exc_value, tb):
                 .format(logging.getLevelName(self.level), self.logger.name))
 
 
+def short_description_from_docstring(text):
```
This function name should start with an underscore since it's not meant to be part of the public API. Also, I think a shorter name should be used like `_make_short_description(docstring)` or `_get_short_description(docstring)`. Information about the arguments can be gotten from the argument names or function docstring.

brettcannon commented:
I talked with our lawyer and he said that as long as Ben and Brian have both signed the CLA then we are fine. So @brianmay can you just tell us your bugs.python.org username and I can manually check you also signed and then this PR will be cleared from a CLA perspective.

Mariatta commented:
I checked that @brianmay has signed the CLA in b.p.o. His b.p.o username is `brian `

brianmay committed:
Change docstring into comment for internal function

brianmay committed:
Merge surrounding space test with existing tests

brianmay committed:
Prefix internal function with _

brianmay committed:
Fix broken docstrings

These docstrings were not PEP 257, and the tests failed with the new
PEP 257 compliant docstring parser.

brianmay commented:
I think I have addressed all feedback. Please let me know if I missed anything. Tests pass on my local box, so hoping this means they will pass the CI checks too.

I have not tried to rebase or merge latest master. If you want me to do so, please let me know.

brianmay commented:
Just noticed I missed updating the docstring for ``shortDescription``, as per feedback. Will do so.

brianmay committed:
Fix docstring for shortDescription

This docstring was truncated in error.

cjerdonek commented on Lib/unittest/case.py:
```diff
@@ -338,6 +339,19 @@ def __exit__(self, exc_type, exc_value, tb):
                 .format(logging.getLevelName(self.level), self.logger.name))
 
 
+def _short_description_from_docstring(text):
```
You should use a shorter name like `_make_short_description(doc)` or `_get_short_description(doc)`. It is unnecessarily verbose to include argument information in the function name, e.g. as this can be obtained from the argument names and/or function docstring. Also, I note that `pydoc`'s convention is to use `doc` for the argument name.

cjerdonek commented on Lib/unittest/case.py:
```diff
@@ -338,6 +339,19 @@ def __exit__(self, exc_type, exc_value, tb):
+def _short_description_from_docstring(text):
+    # Return the summary line from a docstring.
+    # If there is no summary line, return None.
+    if text:
```
It would be cleaner to use a guard clause and just write (and with no comment):

    if not text:
        return None

cjerdonek commented on Lib/unittest/case.py:
```diff
@@ -338,6 +339,19 @@ def __exit__(self, exc_type, exc_value, tb):
 
 
+def _short_description_from_docstring(text):
+    # Return the summary line from a docstring.
```
This should be part of a docstring, not code comment.

serhiy-storchaka replied:
I suggested to replace too verbose docstring in initial commit with shorter comment. But since it was shortened I'm nice with a docstring too.

brianmay committed:
Rename internal function

brianmay committed:
Rename text parameter to doc

brianmay committed:
Use guard clause for internal function

brianmay committed:
Replace comment with docstring

brianmay commented:
I think I have addressed all feedback up to this point.

cjerdonek commented on Lib/unittest/case.py:
```diff
@@ -339,16 +339,17 @@ def __exit__(self, exc_type, exc_value, tb):
+    """
+    Return the summary line from a docstring.
+
+    If there is no summary line, return None.
```
> If there is no summary line, return None.

I don't think this is accurate. If there is no summary line, it can also return the empty string.

cjerdonek commented:
I believe the latest patch results in a behavior change (and an undesirable one). Prior to the patch, a test-case docstring of the form `'abc\ndef'` (i.e. more than one line, and non-empty second line) would yield a short description of `'abc'` (i.e. the first line). With the patch, the short description is `''`. It seems like this should be a test case, too.

cjerdonek commented:
This probably means you can't use `pydoc.splitdoc()`. Otherwise, you will be trading one form of undesirable behavior for another. You probably just want to make `_get_short_description()` call `strip()` before doing what it currently does.

brianmay commented:
I believe a docstring of 'abc\ndef' is not a PEP 257 compliant docstring. Making this PEP 257 compliant was the point of this change.

cjerdonek commented:
The original bug report was only about docstrings that started with whitespace (e.g. a newline):

> The attached ‘test_foo.py’ module contains two test case functions.

Making the parsing PEP 257 compliant is a broader change and would introduce backwards-incompatible changes that I don't think are desirable. For example, cases where users currently get the first line would start getting a less useful empty string.

The reference to PEP 257 was just to point out that it should be okay to start docstrings with a newline -- not to _require_ that everyone follow PEP 257.

vstinner commented:
@brianmay: I'm sorry but our bot is still unable to see that you sign the Python CLA :-( Since you said that you signed it, would you mind to double-check that you entered correctly your GitHub login in your bugs.python.org account? If you filled correctly your bugs.python.org profile, maybe retry to sign the CLA!?

brianmay commented:
@Mariatta said above "I checked that @brianmay has signed the CLA in b.p.o. His b.p.o username is brian".

I don't think there is any need for me to sign the CLA again after it has already been confirmed that I have signed it.

vstinner commented:
> @Mariatta said above "I checked that @brianmay has signed the CLA in b.p.o. His b.p.o username is brian". I don't think there is any need for me to sign the CLA again after it has already been confirmed that I have signed it.

I see that http://bugs.python.org/user14312 has "Contributor Form Received | Yes", bpo login "brian", GitHub login "brianmay".

*But* I just removed the "CLA not signed" label, and the bot ... added it again, as you didn't sign it :-(

I only see one bpo user with github login="brianmay".

Could it be a mismatch based on the mail address?

@brettcannon: Would you mind to take a look please?

Mariatta commented:
@haypo This PR is kinda unique.. What happened is that @brianmay made the PR based on Ben Finney's patch. Because of this the bot tries to check if both Ben and Brian have signed the CLA.
In https://bugs.python.org/msg293522, Ben Finney stated that he does not want to create GitHub account. Since he does not have GitHub account, the bot can't check whether he signed the CLA or not.

I see that both Ben and Brian have signed the CLA. I think from CLA perspective it's clear. 
But wouldn't hurt if both you and/or @brettcannon take another look and confirm that both users have signed the CLA.

Otherwise, If the conflict can be resolved, and if you approve this PR, I think this is good to go.

vstinner commented:
I suggest to export the change as a patch (git diff master.. > patch), create a new PR (git checkout -b pydoc; git apply patch; git commit -a), *and* credit Ben Finney in the commit message, like:

```
Co-Authored-By: Ben Finney <xxx@xxx>
```

cjerdonek commented:
On the PR itself, note that I have review comments from a couple months ago that haven't been addressed.

brettcannon commented:
Both Ben and Brian have signed the CLA and it's just Ben's refusal to set up a GitHub account with his bugs.python.org account that's causing the CLA bot to say this PR isn't in the clear. So as long as there are no other commits from anyone but Ben and Brian (which is what is currently true), then from a CLA perspective we're in the clear in this case.

@cjerdonek can you change your review to "request changes" if you think they should be addressed before this PR is considered for acceptance? Plus there are merge conflicts that have to be fixed.

cjerdonek reviewed (changes requested):
For backwards compatibility reasons, the PR should preserve the behavior that docstrings of the form `'abc\ndef'` return the first line. It should also add a test for this case. While the PR can help ensure that PEP-257 compliant docstrings result in a useful result, it shouldn't require docstrings to be compliant to do so. (Many test-case docstrings out in the wild aren't.)

My suggestion would be to change the implementation of `_get_short_description()` to match the current implementation, with a call to `strip()` beforehand. The docstring of `_get_short_description()` should also be updated accordingly.
